runtime/runtime/src/lib.rs

```diff
@@ -148,6 +147,60 @@ pub struct ApplyState {
     pub bandwidth_requests: BlockBandwidthRequests,
 }
 
+/// Represents a single group of transactions (sharing the same signer_id and public_key)
+/// along with their indices in the original slice.
+pub struct TransactionGroup<'a> {
+    pub indices: Vec<usize>,
+    pub signed_txs: &'a [SignedTransaction],
+}
+
+/// Represents multiple groups of transactions. It provides a parallel iterator over
+/// each individual TransactionGroup.
+pub struct TransactionGroups<'a> {
+    groups: Vec<TransactionGroup<'a>>,
+}
+
+impl<'a> TransactionGroups<'a> {
+    /// Constructs TransactionGroups from a slice of SignedTransaction.
+    /// Transactions are sorted by (signer_id, public_key, nonce) and then grouped by (signer_id, public_key).
+    pub fn new(signed_txs: &'a [SignedTransaction]) -> Self {
+        let mut indices: Vec<usize> = (0..signed_txs.len()).collect();
+        indices.sort_by_key(|&i| {
+            let st = &signed_txs[i];
+            (st.transaction.signer_id(), st.transaction.public_key(), st.transaction.nonce())
+        });
+        let groups = if indices.is_empty() {
+            Vec::new()
+        } else {
+            indices
+                .par_chunk_by(|&left_idx, &right_idx| {
+                    let left = &signed_txs[left_idx];
+                    let right = &signed_txs[right_idx];
+                    left.transaction.signer_id() == right.transaction.signer_id()
+                        && left.transaction.public_key() == right.transaction.public_key()
+                })
+                .filter(|chunk| !chunk.is_empty())
+                .map(|chunk| TransactionGroup { indices: chunk.to_vec(), signed_txs })
+                .collect()
```

Comment:
By collecting here you immediately move out of the `ParallelIterator` context. Move the `par_chunk_by` logic into `par_grouped` and avoid the intermediate work of splitting up groups altogether.